src/rhash/hash.c

```diff
@@ -2407,6 +2407,206 @@ static int rc_hash_gamecube(char hash[33], const rc_hash_iterator_t* iterator)
+  {
+    rc_file_close(iterator, file_handle);
+    return rc_hash_iterator_error(iterator, "Not a Wii disc");
+  }
+
+  /* Check encryption byte - if 0x61 is 0, disc is encrypted */
+  rc_file_seek(iterator, file_handle, 0x61, SEEK_SET);
+  rc_file_read(iterator, file_handle, quad_buffer, 1);
+  encrypted = (quad_buffer[0] == 0);
+
+  /* Hash main headers */
+  buffer = (uint8_t*)malloc(MAIN_HEADER_SIZE);
+  if (!buffer)
+  {
+    rc_file_close(iterator, file_handle);
+    return rc_hash_iterator_error(iterator, "Could not allocate temporary buffer");
+  }
+  rc_file_seek(iterator, file_handle, 0, SEEK_SET);
+  rc_file_read(iterator, file_handle, buffer, MAIN_HEADER_SIZE);
+
+  md5_init(&md5);
+  rc_hash_iterator_verbose_formatted(iterator, "Hashing %u byte main header", MAIN_HEADER_SIZE);
+  md5_append(&md5, buffer, MAIN_HEADER_SIZE);
+  free(buffer);
+
+  /* Scan partition table */
+  rc_file_seek(iterator, file_handle, 0x40000, SEEK_SET);
+  for (ix = 0; ix < 8; ix++)
+  {
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    partition_info_table[ix] =
+      (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+    if (ix % 2 == 0)
+      total_partition_count += partition_info_table[ix];
+  }
+  partition_table = (uint32_t*)malloc(total_partition_count * 4 * 2);
+  kx = 0;
+  for (jx = 0; jx < 8; jx += 2)
+  {
+    for (ix = 0; ix < partition_info_table[jx]; ix++)
+    {
+      rc_file_seek(iterator, file_handle, partition_info_table[jx + 1] << 2, SEEK_SET);
+      rc_file_read(iterator, file_handle, quad_buffer, 4);
+      partition_info_table[kx++] =
+        (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+      rc_file_read(iterator, file_handle, quad_buffer, 4);
+      partition_info_table[kx++] =
+        (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+    }
+  }
+
+  /* If the data is encrypted, all my reads to buffer will be the same size. */
+  /* Might as well allocate now instead of every iteration. */
+  if (encrypted)
+    buffer = (uint8_t*)malloc(CLUSTER_SIZE);
+
+  /* Read each partition */
+  for (jx = 0; jx < (uint32_t)total_partition_count * 2; jx += 2)
+  {
+    /* Don't hash Update and Demo partitions*/
+    if (partition_table[jx + 1] != 0 && partition_table[jx + 1] != 2)
+      continue;
+
+    rc_file_seek(iterator, file_handle, (partition_table[jx] << 2) + 0x2B8, SEEK_SET);
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    part_offset =
+      (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+    rc_file_read(iterator, file_handle, quad_buffer, 4);
+    part_size =
+      (quad_buffer[0] << 24) | (quad_buffer[1] << 16) | (quad_buffer[2] << 8) | quad_buffer[3];
+
+    if (encrypted)
+    {
+      for (ix = 0; ix < part_size / 8 && ix < MAX_CLUSTER_COUNT; ix++)
```

Comment:
I'm not sure where the number 8 came from. I would have thought you should divide by `0x8000 >> 2`.